### ISSUE TITLE
backend feature

### ISSUE THREAD

**Issue by skadefro** [isomorphic-git/isomorphic-git#1531]
Hey, fantastic work!
i know this has been raised before in #672 and #464
Is there any chance of making it possible to support using this module for hosting a git repo as well ?

**Comment by jcubic** [comment]
Only if you start a new project and implement git server.

**Comment by skadefro** [comment]
I know it's mean to be used as a client, but there was some work done on creating the http responses needed in order to use the module for that purpuse.
All git repo examples i've found so far, *all* depend on using the git command using --stateless-rpc
So in way you could say, this is also a git client request ;-)

Considering you already know how to send data, i had hoped it would be simple to also allow us to "create" the responses.
I know some work was done to implement this in the server branch, but parts of that code is now gone and/or not accessible from the current code base. https://github.com/isomorphic-git/isomorphic-git/tree/feat/server

By doing this, i could use my custom "fs" implementation that uses a database instead of the filesystem, to expose a hithub repository. Right now, i need to "dump" the data from the database to the disk on a server, in order to expose a git repo, that is not very effecient.

**Comment by jcubic** [comment]
Unfortunately, the main author and architect of this project is not active anymore. So unless you can create a server yourself and contribute no one will do this for you.

Alternatively, you can hire a freelancer that will implement this.

So I think that the only approach here is for you to fork the repository, figure out how git works, and play with it. Bad thing is that the git server HTTP protocol is probably not documented in a way that will allow you to just follow the docs. So you probably will need to read the git source code in order to implement the server.

And I'm not saying this to discourage you, but I only think it would not be easy and it can take quite some time. But if you want to contribute if we can figure out what the API will look like, I can integrate the change into a project.

**Comment by mojavelinux** [comment]
I think making isomorphic-git a server is way outside the scope of this library. Fortuantely, there's already an excellent git *server* library in the Node.js called [node-git-server](https://yarnpkg.com/package/node-git-server). I use it in testing environments for my projects and it works great.

Granted, that library does delegate to the canonical git command. But it would be possible to integrate isomorphic-git as an alternative git repository manager. I think that codebase provides a better starting point then going at it from this project alone.

**Comment by skadefro** [comment]
I/we depend on the good will of the contributers. I was just hoping since isomorphic-git already know how to talk the protocol, it would be possible to also parse the http requests back to git commands server side. I’ve spent a few days trying to debug it, using some of the git server repos out there, with out any luck. And half my google searches bring me back to  this repo anyway. 
If i could, I would love to write it my self and do a pull request, but this seem a bit over my skills. 
It was worth asking, if this is beyond the scope of this project, please close this issue 😊

**Comment by jcubic** [comment]
@mojavelinux Note that @wmhilton started in this branch with Git server support https://github.com/isomorphic-git/isomorphic-git/compare/feat/server

But if you think that it's out of scope I'm fine. Another idea is to fork [node-git-server](https://yarnpkg.com/package/node-git-server) into the isomorphic-git repo and use isomorphic-git as the backend, @[skadefro](https://github.com/skadefro) if you want to play with something like this. Later we can push the repo to org, also if you want you can keep this project your own. If this is what you want we can only link to your repo from README or FAQ.

**Comment by jcubic** [comment]
I will keep this issue open, and mark it as help wanted if someone will want to implement this.

**Comment by skadefro** [comment]
I would love to help to create something that others can use for hosting git repos, but I need help "cracking the code" on how to parse packages/requests the client's send. If someone can show me in a google meet, or give a short example. Then I'll gladly help continue from there. 
Making "node-git-server" support isomorphic-git makes good sense, but I still think either isomorphic-git should expose something for parsing the packages, or there should be a stand alone npm package for that part ( like pushover have been for most projects out there )

**Comment by Amit89480** [comment]
Is this issue still open as I want to contribute since I am new to open source contribtion

**Comment by jcubic** [comment]
I think that this difficult task to make. You will need to know git internals and how the git server works. I don't know any resources that explain how git servers work. But in [Contribution file](https://github.com/isomorphic-git/isomorphic-git/blob/main/CONTRIBUTING.md#how-git-works) there is a list of resources that explains the internals of git. If you find any article about the server fill free to comment here with the links.

Note that this should not be part of this repository, you can create your own project with git server. But before you start think why you need to have git server? I've never used git server so I'm not sure how to run it, but there are probably already git servers that you can use. You don't need to create another one in JavaScript that would be painfully slow.

So the big question is: why do you want a git server in JavaScript?

**Comment by jcubic** [comment]
Here is instruction on [how to run the git server on GNU/Linux](https://www.linuxfoundation.org/blog/blog/classic-sysadmin-how-to-run-your-own-git-server).

**Comment by jcubic** [comment]
Of course, you can try to create a git server in NodeJS, this would probably be a nice project where you can customize the server to your needs but this is out of scope for isomorphic-git. But if you create a project like this you can move it to the isomorphic-git organization if you want. It can be part of the project but more like lightning-fs and not part of the isomorphic-git repo.

If you want the repo to be in the org from the beginning I can create one for you where you can collaborate. I can also do CR for your changes.

**Comment by jcubic** [comment]
As for helping with parsing git messages, and how to use git internals I can't help I'm afraid. If you want a git server you will need to figure out it on your own. I would start checking the git source code.

**Comment by skadefro** [comment]
I **finally** found some code to "kick start" my implementation of a Git server in Node.js. [cloud-git](https://github.com/fusebit/cloud-git) works quite nicely, but did need some "touches" to work for my case. While testing it, I created an example implementation using MongoDB as a backend (my original goal, by using isomorphic-git) at [cloud-git-mongodb](https://github.com/skadefro/cloud-git-mongodb). I would still prefer a solution based on isomorphic-git, but if no one is interested in sharing more information on how to get started, this will work just fine for my use case.
